### ISSUE TITLE
What data to feed to the UMAP algorithm to get projected 2D graph?

### ISSUE THREAD

**Issue by zhuoqun-chen** [uwgraphics/MotionComparator#24]
Hi @yepw, @zelongjiang123 and the team,

Thank you for releasing the code for your paper!

I'm tring to use `umap-learn` as suggested in your paper to compare some motion trajectories (some are pre-collected and some are generated by controllers). But I currently don't know what is the shape of the matrix data for UMAP algorithm to get similar plots like in your paper.

For a single motion, is it a series of 7 DoF joint angles or also contains velocity readings of each joint? And does all such motions have its label information for each data point? That is, will each row has a column to label whether it belongs to parameter set 1, 2, 3, 4, 5 so on and so forth.

Any help would be appreciated. Thank you!

**Comment by yepw** [comment]
Hi @zhuoqun-chen, thank you for your interest in our work! Are you trying to use the [umap-learn Python package](https://umap-learn.readthedocs.io/en/latest/) or the motion comparator? To clarify, currently motion comparator uses [umap-js](https://github.com/PAIR-code/umap-js), but as mentioned in the paper, we found that umap-learn generally work better.

**Comment by zhuoqun-chen** [comment]
Hi @yepw , I'm trying to use `umap-learn` python package here! Your motion comparator webpage is so amazing and inspires me to analyze our collected trajectories using this UMAP algorithm. I've got some experience visualizing data with t-SNE, however, the results are always cluster-like. And I don't know how to visulize high-dimentional data like robot motion sequences as lines in 2D (demonstrated in your paper).

Would you like to give me some hints on the data format that you feed to `umap-js`? I took a look at the https://github.com/uwgraphics/MotionComparator/blob/main/src/scripts/objects3D/UmapGraph.tsx script, and it is quite similar to the API in https://umap-learn.readthedocs.io/en/latest/.

In the python version, they have an example like this:


```python
import umap
from sklearn.datasets import load_digits

digits = load_digits()

embedding = umap.UMAP(n_neighbors=5,
                      min_dist=0.3,
                      metric='correlation').fit_transform(digits.data)
```

And what is the format of the `digits.data` here in your trails? is it a 2D numpy array where each row represent a configuration at timestep `t` or sth else 😃?

**Comment by yepw** [comment]
In our implementation, it's just a 2D array where each row presents the joint positions at a timestamp. For example, for a 7 Dof robot, there are 7 numbers in a row. Timestamp or label information is not necessary because umap-learn preserves row orders, _i.e._,  the first row of the output 2D array corresponds to the first row of the input 2D array. 
But if you want, include timestamp or joint velocity may produce some interesting results.

**Comment by zhuoqun-chen** [comment]
Thank you for the suggestions~!
